### ISSUE TITLE
[models]: Gemma4 hybrid kv cache is part-disabled and broken

### ISSUE THREAD

**Issue by tchedaTT** [tenstorrent/tt-metal#46432]
### Component / Area

models

### Issue Type (optional)

None

### Observed

https://github.com/tenstorrent/tt-metal/pull/45184 and https://github.com/tenstorrent/tt-metal/pull/45404 disabled parts of the hybrid attention - kv cache machinery. As acknowledged by the first PR, gemma4 overrides get_kv_cache_spec and keeps emitting SlidingWindowSpec.
This leads to inconsistencies. For example, the modifications made warmup no longer cover per-layer tables. As reported by a user, this can lead to bad outputs.

### Expected

Either fully enable or fully disable the machinery. Ensure proper warmup.

### 1. Steps (exact commands)

-

### 2. Input data / link or description

-

### 3. Frequency

-

### 1. Software Versions

-

### 2. Hardware Details

-

### Is this a regression?

Yes

### Priority

None

**Comment by tchedaTT** [comment]
cc @handrewsTT @arginugaTT

**Comment by sjalloq** [comment]
Hi @tchedaTT,

Here's a more detailed report - I've pushed everything to my [play repo](https://github.com/sjalloq/tt-play/tree/main) with links to the relevant docs at the bottom of this ticket.  Full disclosure - I've been using Claude and that repo as a learning tool.  No experience here previously.

### Summary
  google/gemma-4-12B-it (text-only, single P150, 1×1) emits multilingual garbage from the first token when served via vllm serve (TT plugin), while offline parity is PCC 1.0
  (prefill, decode, traced decode). The cause is in the gemma4 vLLM bridge's decode warmup — not the model, and not the kv-cache-groups layout. One-line-ish fix; keeps decode
  tracing (full speed).

###  Environment
  tt-metal arg/gemma4_optimizations (+f7d0161); tenstorrent/vllm 5eb61e8 + TT plugin; single P150, MESH_DEVICE=P150, blackhole-1x1; serve --max-num-seqs 1 --max-model-len 4096
  --block-size 64 --no-enable-prefix-caching, host sampling, trace_mode: all.

###  Symptom
  /v1/completions "The capital of France is" → garbage ("a canonical droit deedUF …"), greedy, from token 1. Offline test_vllm_parity.py → PCC 1.0.

###  Root cause
  Gemma4ForCausalLM doesn't override warmup_model_decode. The inherited warmup captures the decode trace while _persistent_per_layer_page_tables don't yet exist (they're allocated
  lazily on the first request). So:
  1. At capture, _active_page_tables_per_layer is None → Gemma4Model.ttnn_decode_forward falls back to the legacy single-page_table path → the trace binds the legacy page_table,
  not the hybrid per-layer buffers.
  2. The 48 per-layer buffers are then allocated after the trace is live (allocator.cpp:110 "unsafe allocation").
  The warmup trace is reused every step, so hybrid full-attention layers address the wrong KV → garbage.

###  Confirmed:
  - trace_mode: none → coherent; trace_mode: decode_only (eager prefill, traced decode) → garbage ⇒ it's the decode trace.
  - Instrumented ordered log: ALLOC _persistent_per_layer_page_tables (n=48) fires after begin/end_trace_capture.

###  Fix — models/demos/gemma4/tt/generator_vllm.py
  Override Gemma4ForCausalLM.warmup_model_decode to pre-allocate the per-layer buffers and set _active_page_tables_per_layer before super().warmup_model_decode captures the trace.
  Idempotent; no speed loss. (~30 lines; PR attached.)
  Validated: instrumented trace_mode: all re-run → ALLOC now precedes capture; both endpoints return "…Paris." at full traced speed (~17.9 tok/s/user).

  Not the same as #45763 / #45789
  kv_cache is stable across steps (vLLM passes the same runner.kv_caches), and allocator.cpp:110 still fires after the fix while output is correct — so this is not the
  kv_cache-ownership / trace-memory-safety issue. #45789's identity guard wouldn't have caught it; this is complementary. Follow-up to the merged bridge #44265 (@handrewsTT).

###  Caveats
  - No offline regression test reproduces it (the warmup_then_inference unit test re-captures the trace, so it passes either way); needs a live serve.
  - Sibling variants (E2B / 26B-A4B / 31B) share this warmup path — validate no regression.
  - Residual allocator.cpp:110 warning remains (benign).

  Full diagnosis trail: [vllm-12b-serving-bug.md](https://github.com/sjalloq/tt-play/blob/main/docs/source/porting/models/gemma4-12b/vllm-12b-serving-bug.md) (bug+fix) and [HYPOTHESIS.md](https://github.com/sjalloq/tt-play/blob/main/docs/source/porting/models/gemma4-12b/HYPOTHESIS.md) (localization log, incl. the ruled-out
  kv-cache-groups / HMA-tensor-sharing layout hypotheses). Repro scripts in demos/gemma4/.

### tt-metal fix
```shell
git diff models/demos/gemma4/tt/generator_vllm.py
diff --git i/models/demos/gemma4/tt/generator_vllm.py w/models/demos/gemma4/tt/generator_vllm.py
index ec493597acf..4d762ea4c0f 100644
--- i/models/demos/gemma4/tt/generator_vllm.py
+++ w/models/demos/gemma4/tt/generator_vllm.py
@@ -415,6 +415,41 @@ class Gemma4ForCausalLM(HybridAttentionForCausalLM):
             # the same skip pattern as the GptOssForCausalLM sibling.
             return super(HybridAttentionForCausalLM, self).decode_forward(*args, **kwargs)

+    def warmup_model_decode(self, *args, **kwargs):
+        """Pre-allocate the per-layer page-table buffers and bind the per-layer
+        routing *before* the decode trace is captured.
+
+        Root cause of the 12B vLLM serving garbage: the inherited warmup
+        captures the decode trace while ``_persistent_per_layer_page_tables``
+        does not yet exist (they're created lazily on the first real request).
+        So the captured trace binds the *legacy* single-page_table path, and
+        the per-layer buffers are then allocated *after* the trace is live
+        (``allocator.cpp:110`` "unsafe allocation"), aliasing trace memory.
+        The reused trace reads the legacy binding → hybrid full-attention
+        layers address the wrong KV blocks → multilingual garbage from token 1.
+
+        Allocating the persistent buffers here (pre-capture, pre-trace) makes
+        the warmup trace bind the per-layer routing path and removes the
+        post-capture allocation. Idempotent: a second warmup call finds the
+        buffers already present.
+        """
+        import torch
+
+        num_blocks = kwargs.get("num_blocks")
+        if num_blocks is None and len(args) >= 4:
+            num_blocks = args[3]
+        if num_blocks is not None:
+            for m in self.model:
+                warmup_pt = [torch.zeros(1, int(num_blocks), dtype=torch.int32) for _ in range(len(m.layers))]
+                m._page_tables_to_ttnn(warmup_pt)  # allocate persistent buffers while no trace is active
+                m._active_page_tables_per_layer = warmup_pt
+        try:
+            return super().warmup_model_decode(*args, **kwargs)
+        finally:
+            for m in self.model:
+                if hasattr(m, "_active_page_tables_per_layer"):
+                    del m._active_page_tables_per_layer
+
     def allocate_kv_cache(self, *args, **kwargs):
         # Legacy uniform path (vLLM falls back here when ``get_kv_cache_spec``
         # isn't consulted). The hybrid path uses ``allocate_kv_cache_per_layer``
```
